tools/nstar/src/main.rs

```diff
@@ -342,6 +355,61 @@ async fn main() -> Result<()> {
                 process::exit(0);
             }
         }
+        // Start process inside running container and attach a pseudo-terminal to it
+        Subcommand::Exec { name, version, cmd } => {
+            // Connect
+            let mut client = Client::new(io, None, opt.timeout)
+                .await
+                .context("Failed to connect")?;
+
+            let version = or_version(&name, version, &mut client).await?;
+            let cmd: Result<Vec<_>> = cmd
+                .into_iter()
+                .map(|c| NonNullString::try_from(c).context("Invalid arg"))
+                .collect();
+
+            // TODO Avoid the nix dependency by opening the PTY in the runtime and sending the
+            // master FD through the client socket.
+            let master_fd = pty::posix_openpt(fcntl::OFlag::O_RDWR | fcntl::OFlag::O_NONBLOCK)?;
+            pty::grantpt(&master_fd)?;
+            pty::unlockpt(&master_fd)?;
+            let pty = unsafe { pty::ptsname(&master_fd) }?;
+
+            let request = Request::Exec {
+                container: Container::new(name.try_into()?, version),
+                cmd: cmd?,
+                pty: PathBuf::from(pty),
+            };
+
+            client
+                .request(request)
+                .await
+                .context("Failed to send request")?;
+
+            // TODO refactor using tokio
+            let master = master_fd.as_raw_fd();
+            loop {
+                let mut fd_set = select::FdSet::new();
+                fd_set.insert(nix::libc::STDIN_FILENO);
+                fd_set.insert(master);
+
+                let ret = select::select(None, &mut fd_set, None, None, None).unwrap();
+                nix::errno::Errno::result(ret).unwrap();
+
+                if fd_set.contains(nix::libc::STDIN_FILENO) {
+                    let mut buf = [0; 512];
+                    let n = unistd::read(nix::libc::STDIN_FILENO, &mut buf).unwrap();
+                    unistd::write(master, &buf[..n]).unwrap();
+                }
+
+                if fd_set.contains(master) {
+                    let mut buf = [0; 512];
+                    let n = unistd::read(master, &mut buf).unwrap();
+                    unistd::write(nix::libc::STDOUT_FILENO, &buf[..n]).unwrap();
+                }
+            }
```

Comment:
ptys and Tokio behave special. I tried to use it the straight forward way in the runtime for the output to log forwarding. Not sure yet what's the specialties are here.